### PR TITLE
fix issue 18149 - Add a compiler trait to detect if a function is @disable

### PR DESCRIPTION
**Pull request by ghost**
Although there's not much concrete application (since a call to a `@disable` function is already statically checked) for now there is no way to know if a function is `@disable`.

### CODE REVIEW

**Comment by JinShil** [comment]
Is there any particular reason why `@disabled` can't be added to the `getAttributes` trait?

**Comment by BBasile** [comment]
> Adding @disable to getAttributes would introduce a breaking change (and may not even be 

Yes. This and also it'd be for the `getFunctionAttributes` verb (the one you ask for is  for the UDAs). There's a lot of code in `std.typecons` that relies on them and `@disable` would not being usable in the same fashion anyway (typically all the wrappers things).

**Comment by JinShil** [comment]
I'm itching to approve this, but before then it needs a [changelog entry](https://github.com/dlang/dmd/tree/master/changelog) in this PR and a documentation update at https://github.com/dlang/dlang.org.

**Comment by BBasile** [comment]
Yes, I was waiting for the next  cleanup of the folder containing the entries. Probably only a matters of days.

**Comment by BBasile** [comment]
@JinShil, a changelog entry is not needed. An issue is referenced so it would appear even without.

**Comment by JinShil** [comment]
> @JinShil, a changelog entry is not needed. An issue is referenced so it would appear even without.

Yes, I know, and I generally agree, but I've been asked to add changelog entries for certain Bugzilla issues, even though they appear in list of fixed issues.  The following are examples:

1.  https://dlang.org/changelog/2.078.0.html#fix16997
2.  https://dlang.org/changelog/2.078.0.html#fix11006

This is a new feature and I think it is appropriate to do some advertising for it.

**Comment by ZombineDev** [comment]
We have had examples of both adding a dedicated changelog entry and relying on Bugzilla (e.g. https://dlang.org/changelog/2.078.0.html#bugfix-list - `isFuture`). IMO we should add changelog entries for all enhancement requests which allow expressing new things in the language, that were not easily doable before.

@RazvanN7 can you add a changelog entry for `isFuture`?

**Comment by BBasile** [comment]
Okay i 've just added the changelog entry and the good news is that this has revealed that the trait is much more useful than i thought.

**Review by JinShil — APPROVED** [review]
Thanks @BBasile.  I'm going to leave this open for a little while to give others an opportunity to weigh in.

**Comment by wilzbach** [comment]
> I'm going to leave this open for a little while to give others an opportunity to weigh in.

Yes this is a language change and thus needs approval from either @WalterBright or @andralex.

File: src/dmd/declaration.d
Comment (WalterBright) [inline src/dmd/declaration.d]: Add `const pure nothrow @nogc @safe`

File: changelog/disabled-trait.dd
Comment (WalterBright) [inline changelog/disabled-trait.dd]: Remove `your`

**Review by WalterBright — CHANGES_REQUESTED** [review]
see my comments

**Comment by BBasile** [comment]
Okay done. Note that I still think that the changelog entry is a bit too much. It's really a minor addition.

**Comment by JinShil** [comment]
@WalterBright Are you willing to explicitly mark this "Approved"?

**Comment by JinShil** [comment]
@BBasile Please rebase.  It won't elicit any reviews unless it's passing.

**Comment by wilzbach** [comment]
> @BBasile Please rebase. It won't elicit any reviews unless it's passing.

What do you mean by this? I don't see any merge conflicts and only auto-tester has one spurious failure, but auto-tester automatically invalidates and restarts all jobs if at any repo a PR is merged to master.

@WalterBright needs to dismiss his review (I could do that as well, but I'm not sure he explicitly approved the addition).

**Comment by JinShil** [comment]
> What do you mean by this?

It appears this code is using `STCdisabled` instead of the recently refactored `STC.disabled`.  See the failure.  That is why I assumed it needed to be rebased.

**Comment by wilzbach** [comment]
> It appears this code is using STCdisabled instead of the recently refactored STC.disabled. See the failure. That is why I assumed it needed to be rebased.

Ah I see. Though we can easily do that and reduce the latency / "ping-pong" time ;-)

**Comment by wilzbach** [comment]
(I just rebased this PR)

**Comment by BBasile** [comment]
Looks like there's a regression under OSX 32 bits, that happened between the 2018-01-10 22:21:41 and the 2018-01-10.

**Comment by JackStouffer** [comment]
>Looks like there's a regression under OSX 32 bits

Can we please stop spending precious dev time and stalling PRs to support this dead platform?

**Comment by wilzbach** [comment]
Oh that's interesting. It looks similar to what @marler8997 is experiencing.
Yeah we definitely should switch that machine to do `64_64` testing only. CC @braddr 
Alternatively we can also try to ping Brad here: https://github.com/braddr/d-tester/issues

**Comment by marler8997** [comment]
Looking at the logs it is the same issue.  I was able to reproduce on my macbook pro and got a stack trace.  Note that this seg fault only occurs because the autotester is using dmd version 2.068.2, I was able to fix the problem on my macbook by using the next version, 2.069.0.

Here's the stack trace, maybe we should consider patching dmd 2.068.2?
```
Obj::term(char const*) + 2285
obj_end(Library*, File*) + 37
tryMain(unsigned long, char const**) + 12066
_start + 203
start + 33
```

**Comment by marler8997** [comment]
P.S. I worked around the issue by adding some printf statements to a certain location in the code :)  It was hours of painstaking trial and error to massage the code to not cause the segfault.  Tedious work, but no one else is going to do it for me.

**Comment by braddr** [comment]
reminder: I do not define the supported platforms.  I only implement them and remind people when they want them changed for whatever reason that those changes must go through the community and leadership team.  I'm part of the former but not the latter.  So, go, get approval for changing the supported platform set.

**Comment by BBasile** [comment]
@wilzbach, circle ci here too doesn't work anymore, although [it used to](https://circleci.com/gh/dlang/dmd/tree/pull%2F7569). I did rebase or amend several times yesterday and it's not triggered anymore (as for semaphore).

**Comment by wilzbach** [comment]
> circle ci here too doesn't work anymore

Interesting that that only happens with your account, but it seems like closing & reopening worked here...

**Comment by BBasile** [comment]
Recently i revoked circle-ci from my account...but i don't think this could be the cause.

**Comment by wilzbach** [comment]
> but i don't think this could be the cause.

It could also be related to the partial GitHub outage a couple of days ago:

https://status.github.com/messages

**Comment by BBasile** [comment]
@WalterBright, you took a look at it 10 days ago, i've adressed your review comments. Can you take a decision quickly ?

**Review by andralex — APPROVED** [review]

**Pull request merged**
